crates/ty_python_semantic/resources/mdtest/typed_dict.md

```diff
@@ -2019,5 +2019,129 @@ static_assert(is_disjoint_from(TD, dict[str, int]))  # error: [static-assert-err
+
+```py
+from typing import TypedDict, Literal
+
+class Foo(TypedDict):
+    tag: Literal["foo"]
+
+class Bar(TypedDict):
+    tag: Literal[42]
+
+class Baz(TypedDict):
+    tag: Literal["baz"]
+
+def _(u: Foo | Bar | Baz):
+    if u["tag"] == "foo":
+        reveal_type(u)  # revealed: Foo
+    elif u["tag"] == 42:
+        reveal_type(u)  # revealed: Bar
+    else:
+        reveal_type(u)  # revealed: Baz
+```
+
+We can descend into intersections to discover `TypedDict` types that need narrowing:
+
+```py
+from collections.abc import Mapping
+from ty_extensions import Intersection
+
+def _(u: Foo | Intersection[Bar, Mapping[str, int]]):
+    if u["tag"] == "foo":
+        reveal_type(u)  # revealed: Foo
+    else:
+        reveal_type(u)  # revealed: Bar & Mapping[str, int]
+```
+
+We can also narrow a single `TypedDict` type to `Never`:
+
+```py
+def _(u: Foo):
+    if u["tag"] == "foo":
+        reveal_type(u)  # revealed: Foo
+    else:
+        reveal_type(u)  # revealed: Never
+```
+
+Narrowing is restricted to `Literal` tags, though, because `x == "foo"` doesn't generally tell us
+anything about the type of `x`. Here's an example where narrowing would be tempting but unsound:
+
+```py
+from ty_extensions import is_assignable_to, static_assert
+
+class NonLiteralTD(TypedDict):
+    tag: int
+
+def _(u: Foo | NonLiteralTD):
+    if u["tag"] == "foo":
+        # We can't narrow the union here...
+        reveal_type(u)  # revealed: Foo | NonLiteralTD
+    else:
+        # ...(even though we can here)...
+        reveal_type(u)  # revealed: NonLiteralTD
+
+# ...because `NonLiteralTD["tag"]` could be assigned to with one of these, which would make the
+# first condition above true at runtime!
+class WackyInt(int):
+    def __eq__(self, other):
+        return True
+
+_: NonLiteralTD = {"tag": WackyInt(99)}  # allowed
+```
+
+We can still narrow `Literal` tags even when non-`TypedDict` types are present in the union:
+
+```py
+def _(u: Foo | Bar | dict):
+    if u["tag"] == "foo":
+        # TODO: `dict & ~<TypedDict ...>` should simplify to `dict` here, but that's currently a
+        # false negative in `is_disjoint_impl`.
+        reveal_type(u)  # revealed: Foo | (dict[Unknown, Unknown] & ~<TypedDict with items 'tag'>)
+
+# The negation(s) will simplify out if we add something to the union that doesn't inherit from
+# `dict`. It just needs to support indexing with a string key.
+class NotADict:
+    def __getitem__(self, key): ...
+
+def _(u: Foo | Bar | NotADict):
+    if u["tag"] == 42:
+        reveal_type(u)  # revealed: Bar | NotADict
+```
+
+It would be nice if we could also narrow `TypedDict` unions by checking whether a key (which only
+shows up in a subset of the union members) is present, but that isn't generally correct, because
+"extra items" are allowed by default. For example, even though `Bar` here doesn't define a `"foo"`
+field, it could be *assigned to* with another `TypedDict` that does:
```

Comment:
I guess that means that it _would_ be sound to narrow using `not in` in some situations...? (But obviously not something for this PR, it can wait for a followup)

```
from typing import TypedDict

class Foo(TypedDict):
    x: int

class Bar(TypedDict):
    y: int

def f(td: Foo | Bar):
    if "x" not in td:
        reveal_type(td)  # we can safely narrow to `Bar` here,
                         # -- the type cannot be `Foo`, since `x` is a required
                         # key in `Foo`
```